contributors/design-proposals/storage/csi-snapshot.md

```diff
@@ -0,0 +1,364 @@
+	// This value may not be empty.
+	Snapshotter string
+
+	// Parameters holds parameters for the snapshotter.
+        // These values are opaque to the system and are passed directly
+        // to the snapshotter. The only validation done on keys is that they are
+        // not empty. The maximum number of parameters is
+        // 512, with a cumulative max size of 256K
+        // +optional
+	Parameters map[string]string
+}
+
+
+```
+### Volume API Changes
+
+With Snapshot API avaialble, users could provision volumes from snapshot and data will be prepopulated to the volumes. Also considering clone and other possible storage operations, there could be many different types of sources used for populating the data to the volumes. In this proposal, we add a general "DataSource" which could be used to represent different types of data sources.
+
+#### The `DataSource` Object in PVC
+
+Add a new `DataSource` field into both PVC and PV to represent the source of the data which is prepopulated to the provisioned volume. If an external-provisioner does not understand the new `DataSource` field and does not pre-populate the new volume, PV/PVC controller should be able to detect that (e.g. PVC has `DataSource` but PV does not) and fail the operation.
+
+Possible `DataSource` types may include the following:
+
+    * VolumeSnapshot: restore snapshot to a new volume
+    * PersistentVolumeClaim: clone volume which is represented by PVC
+
+```
+
+type PersistentVolumeClaimSpec struct {
+        // If specified, volume will be prepopulated with data from the PVCDataSourceRef.
+        // +optional
+        DataSourceRef *TypedLocalObjectReference `json:"dataSourceRef" protobuf:"bytes,2,opt,name=dataSourceRef"`
+}
+
+type PersistentVolumeSpec struct {
+        // If specified, volume will be prepopulated with data from the PVCDataSourceRef.
+        // +optional
+        DataSourceRef *ypedLocalObjectReference `json:"dataSourceRef" protobuf:"bytes,2,opt,name=dataSourceRef"`
+}
+
+```
+
+Add a TypedLocalObjectReference in core API.
+
+```
+
+// TypedLocalObjectReference contains enough information to let you locate the referenced object inside the same namespace.
+type TypedLocalObjectReference struct {
+        // Name of the object reference.
+        Name string
+	// Kind indicates the type of the object reference.
+	Kind string
+}
+
+```
+
+In the first version, only VolumeSnapshot is the supported `Type` for data source object reference. Other types will be added in a future version. If unsupported `Type` is used, the PV Controller SHALL fail the operation.
+
+
+#### Add `DataSource` to `StorageClass`
+
+DataSource will be added as an optional parameter in `StorageClass` to allow volumes belonging to the same `StorageClass` to be created with prepopulated data.
+
+```
+
+type StorageClass struct {
+        // If specified, volume will be prepouplated with data from the DataSource
+        // +optional
+        DataSource TypedLocalObjectReference
```

Comment:
Do we really need this? How common is it?

What if this and the PVC both specify source, and they disagree?

Reply:
For all of these source fields - what happens if I point them at a nonsensical kind, like "Service" or at a kind that doesn't exist?  What if I point them at a valid Kind but the named instance doesn't exist.  they ALL need to document this.